### PR TITLE
Hack week: Update target SDK from 31 to 33

### PR DESCRIPTION
**Pull request by 0nko**
This PR updates the target SDK from 31 to 33 and makes the [necessary changes](https://developer.android.com/about/versions/13/behavior-changes-13):

- [Moved the package namespace declarations from the manifests to gradle files](https://github.com/woocommerce/woocommerce-android/commit/e95470306749361bac7474d544ad312226216321)
- ~~[Updated Gradle version to 7.3.0, which addressed the warning caused by the target SDK update](https://github.com/woocommerce/woocommerce-android/commit/2ee0173cf78bae8fc867b513fead78a78ab336de)~~
- Implemented a notification permission request bar, which is displayed now if the app does not have the permission to send push notifications. The users can dismiss the bar and they won't be bothered by it afterwards. This changes the current behavior, which displayed the permission request on devices running Android 13 (API 33) immediately after the app launch. [More info here](https://developer.android.com/develop/ui/views/notifications/notification-permission).

| Request bar | Permission dialog |
| --- | --- |
| image | image |

[More changes are required](https://developer.android.com/about/versions/13/behavior-changes-13#granular-media-permissions) in the Media library, which are addressed in a [separate PR](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/63).

**To test:**

You'll need a device running Android version 13 (API level 33) to test the new bahavior.

1. Start the app with a clean installation
2. Notice the notifications permission request is not displayed immediately after the launch anymore
3. Proceed with the login
4. Notice that a banner pops up at the bottom with a description of the permission request
5. Tap on Allow
6. Grant the notifications permission
7. Kill the app and start it again
8. Notice the banner doesn't pop up
9. Go to App info -> Permissions
10. Verify that `Notifications` is listed in the `Allowed` section
11. Go to the Storage settings and clear the app storage
12. Go back to the app and log in
13. Notice the permission banner pops up again
14. Tap on Dismiss
15. Kill the app and start it again
16. Notice the app doesn't pop up
17. Go to App info -> Permissions
18. Verify that `Notifications` is listed in the `Not Allowed` section

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -1,5 +1,5 @@
 pluginManagement {
-    gradle.ext.agpVersion = '7.2.1'
+    gradle.ext.agpVersion = '7.3.0'
```
Comment (ParaskP7) [inline settings.gradle]: 👋 @0nko !

Just an FYI that updating AGP would (most probably) still require a coordinated effort because of [composite builds](https://github.com/woocommerce/woocommerce-android/blob/trunk/local-builds.gradle-example), just like we've been doing for a few Gradle & AGP versions already (example project: `paaHJt-2VL-p2` + last such updates [here](https://github.com/woocommerce/woocommerce-android/pull/5755) and [here](https://github.com/woocommerce/woocommerce-android/pull/6824)).

PS: This also means that WPAndroid and all its dependant [composite build](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/local-builds.gradle-example) related libraries would need to be updated at the same time as well (example of last such WPAndroid update [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16834), which corresponds to the above WCAndroid [here](https://github.com/woocommerce/woocommerce-android/pull/6824)).

Comment (0nko) [reply settings.gradle]: Hey @ParaskP7! 👋 Thanks for the heads up.. Hmm, maybe we could keep the old AGP version and do it separately, then. I didn't realize it'd cause that much trouble. I only updated it because there was a warning, however, the build worked fine.

Comment (ParaskP7) [reply settings.gradle]: 👋 @0nko !

> Hmm, maybe we could keep the old AGP version and do it separately, then.

Yes, I think that would be for the best. Doing this AGP version update now might add some extra (unintended) work.

> I didn't realize it'd cause that much trouble. I only updated it because there was a warning, however, the build worked fine.

Yes, that's true, I've been observing these warnings for some time too. FYI: Those warnings started to appear when the `compileSdkVersion` got updated to `33`.

**Comment by 0nko** [comment]
@kidinov @JorgeMucientes @ParaskP7 Because you like this PR so much, you'll be happy to know it's ready for a review 😝

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt [hunk 2]
```diff
@@ -54,6 +59,14 @@ class MainPresenter @Inject constructor(
...
     private var pendingUnfilledOrderCountCheck: Boolean = false
 
+    private val shouldShowNotificationsPermissionBar: Boolean
+        get() {
+            return VERSION.SDK_INT >= VERSION_CODES.TIRAMISU &&
+                mainView?.hasNotificationsPermission == false &&
+                !AppPrefs.getWasNotificationsPermissionBarDismissed() &&
+                selectedSite.get().connectionType == Jetpack
+        }
+
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt]: Minor np, I'd say we tend to add new UI logic to `MainActivityViewModel` for the sake of getting rid of the few presenters left bit by bit. Furthermore, adding it to the view model should be more straightforward as you won't have to add any interface functions and its implementations,  just expose the  `NotificationsPermissionBar` visibility as live data.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/message/NotificationsPermissionBarView.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import androidx.activity.result.contract.ActivityResultContracts.RequestPermission
+import androidx.appcompat.app.AppCompatActivity
+import com.google.android.material.card.MaterialCardView
+import com.woocommerce.android.AppPrefs
+import com.woocommerce.android.R
+import com.woocommerce.android.databinding.NotificationsPermissionsBarBinding
+import com.woocommerce.android.util.WooAnimUtils
+import com.woocommerce.android.util.WooPermissionUtils
+
+class NotificationsPermissionBarView @JvmOverloads constructor(
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/message/NotificationsPermissionBarView.kt]: Curious why didn't you go the Compose way here? Was it less straightforward?

File: WooCommerce/src/main/res/layout/fragment_my_store.xml [hunk 4]
```diff
@@ -40,10 +41,10 @@
 
                     <androidx.compose.ui.platform.ComposeView
                         android:id="@+id/store_onboarding_view"
+                        style="@style/Woo.Card"
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/res/layout/fragment_my_store.xml]: Interesting, why did you decide to add this style?  Haven't noticed any impact on the UI after this change. Also, this style change will impact (in case it has an visible impact) the onboarding full screen mode which is not a card UI. 
I'd say that if we need to fix any UI glitch on the onboarding UI, I'd handle it in the compose code. Let me know wdyt.

File: libs/cardreader/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.woocommerce.android.cardreader">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
     <uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
     <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
```
Comment (JorgeMucientes) [inline libs/cardreader/src/main/AndroidManifest.xml:7]: Is this related to targeting API 33? Why are both needed now?

File: WooCommerce/src/main/res/layout/notifications_permissions_bar.xml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            app:layout_constraintTop_toBottomOf="@id/notifications_permission_title" />
+
+        <!-- PRIMARY button -->
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/btn_dismiss"
+            style="@style/Woo.Card.Button"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/dismiss"
+            android:layout_margin="0dp"
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/res/layout/notifications_permissions_bar.xml]: Np. Needed?

File: WooCommerce/src/main/res/layout/notifications_permissions_bar.xml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            app:layout_constraintTop_toBottomOf="@+id/notifications_permission_description" />
+
+        <!-- SECONDARY button -->
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/btn_allow"
+            style="@style/Woo.Card.Button"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/allow"
+            android:layout_margin="0dp"
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/res/layout/notifications_permissions_bar.xml]: Np. Needed?

**Review by JorgeMucientes — APPROVED** [review]

File: WooCommerce/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -2251,6 +2253,8 @@
     <string name="permissions_denied_message">It looks like you turned off permissions required for this feature.&lt;br/&gt;&lt;br/&gt;To change this, edit your permissions and make sure &lt;strong&gt;%s&lt;/strong&gt; is enabled.</string>
     <string name="permission_camera">Camera</string>
     <string name="button_edit_permissions">Edit permissions</string>
+    <string name="notifications_permission_title">Notifications</string>
+    <string name="notifications_permission_description">We need your permission to send you push notifications for new orders, reviews, etc. delivered to your device.</string>
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/res/values/strings.xml:2260]: Is the string value correct? Feels a bit weird after the `etc` part. Should it be a comma after `etc`? Although I'm not an expert in English 😅

**Comment by JorgeMucientes** [comment]
Thanks for handling this task @0nko! 🏅. I did a first round and didn't find anything blocking. But I left a couple of questions and a few nitpicks I think are worth considering looking into before merging. 

Also, I've seen there are some changes related to media picker library. Is there anything I should test related to that?

File: libs/cardreader/src/main/AndroidManifest.xml (hunk 5 above)
Comment (0nko) [reply libs/cardreader/src/main/AndroidManifest.xml:7]: I'm not sure if it was caused by API 33 but I started seeing this warning, so I added this.

image

File: WooCommerce/src/main/res/layout/notifications_permissions_bar.xml (hunk 6 above)
Comment (0nko) [reply WooCommerce/src/main/res/layout/notifications_permissions_bar.xml]: I added this so that the button text would have the same margin as the rest of the layout and IMO it looks better, but I can get rid of it.

image
image

Without the margin:
image

File: WooCommerce/src/main/res/values/strings.xml (hunk 8 above)
Comment (0nko) [reply WooCommerce/src/main/res/values/strings.xml:2260]: I just copied it from `jetpack_benefits_modal_push_notifications_subtitle` (`Get push notifications for new orders, reviews, etc. delivered to your device.`) and modified it, so I didn't really think about it. But [it is ok](https://www.editage.com/insights/research-paper-comma-after-etc-at-the-end-of-a-sentence#:~:text=Generally%2C%20in%20American%20English%2C%20if,as%20the%20final%20punctuation%20mark.) :).

**Comment by 0nko** [comment]
Thanks for the review @JorgeMucientes, ready for another round.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt (hunk 2 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt]: Oh, I wasn't even aware we had `MainActivityViewModel` since there was still a presenter 😅. I'll move it, thanks!

File: WooCommerce/src/main/res/layout/notifications_permissions_bar.xml (hunk 6 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/res/layout/notifications_permissions_bar.xml]: Yeah I agree looks better. Let's keep it as it is 👍🏼

File: WooCommerce/src/main/res/layout/fragment_my_store.xml (hunk 4 above)
Comment (0nko) [reply WooCommerce/src/main/res/layout/fragment_my_store.xml]: There was some UI inconsistency with the shadow or margins but I don't see it anymore.. I'll just get rid of it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/message/NotificationsPermissionBarView.kt (hunk 3 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/message/NotificationsPermissionBarView.kt]: It was just easier with the presenter, but now that I know about the VM, I converted it. I had to change the UI a bit because I was having trouble showing a transparent background behind the Card with `ComposeView`. Let me know what you think.

image

**Comment by 0nko** [comment]
Thanks for the review, @JorgeMucientes. I've addressed your comments, ready for another round.

**Comment by 0nko** [comment]
Btw, the Media Picker has been already tested separately, so you don't need to spend too much time on it. Smoke testing is fine.

**Comment by JorgeMucientes** [comment]
Thanks for applying the suggestions @0nko. The code looks good and works as expected. :shipit:
